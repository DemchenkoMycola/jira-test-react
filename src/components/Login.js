import React from 'react';
import {connect} from 'react-redux';
import { Redirect } from 'react-router-dom';

class Login extends React.Component{
    constructor(props){
        super(props);
        this.state = {
            name: '',
            password: '',
            login: false,
            user: {},
            errors: '',
        }

        this.handleLogin = this.handleLogin.bind(this);
        this.handleName = this.handleName.bind(this);
        this.handlePassword = this.handlePassword.bind(this);
        this.handleUser = this.handleUser.bind(this);
    }

    handleUser(name){
        let  users  = this.props.users;
        let user = users.find((user) => 
            user.name === name || user.email === name
        )
        
        if(user){
            // console.log(user)
            this.setState({
                user: user
            })
        }

    }

    handleLogin(){

            if(this.state.user.password === this.state.password){
                this.setState({
                    login: true
                })
            }else{
                this.setState({
                    errors: "incorect username or password"
                })
            }
    }

    handleName(event){
        let name = event.target.value
        this.setState({
            name: name
        })
        this.handleUser(name);
    }

    handlePassword(event){
        this.setState({
            password: event.target.value
        })
    }

    render(){
        if(this.state.login){
            return <Redirect to={`/projects/${this.state.user.id}`} />
        }
        return(
            <div>
                <section className="login_form">
                    <legend>Login</legend>
                    <div className="errors">{this.state.errors}</div>
                    <label>Name</label>
                    <input 
                        type="text" 
                        placeholder="name"
                        onChange={this.handleName}
                        value={this.state.name}
                        />
                    <label>Password</label>                    
                    <input 
                        type="password" 
                        placeholder="password" 
                        onChange={this.handlePassword}
                        value={this.state.password}
                        />
                    <button 
                    className="button" 
                    type="submit"
                    disabled={!this.state.name && !this.state.password}
                    onClick={this.handleLogin}
                    >Login</button>
                </section>
            </div>
        )
    }
}

const mapStateToProps = (state) => {
    return{
        users: state.users
    }
}


export default connect(mapStateToProps)(Login);