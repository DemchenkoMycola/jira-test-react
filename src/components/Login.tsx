import React from 'react';
import {connect} from 'react-redux';
import { Redirect } from 'react-router-dom';
import { loginUser } from '../actions/userActions';

// interface User {
//     name: string,
//     password: string
// }

interface Props{
    users: object[],
    loginUser: (user: any) => void
}

interface State {
    name: string,
    password: string,
    user: any,
    errors: string,
    login: boolean
}

type InputEvent = React.ChangeEvent<HTMLInputElement>;

class Login extends React.Component<Props, State>{
    constructor(props: any){
        super(props);
        this.state = {
            name: '',
            password: '',
            user: {
                name: '',
                password: ''
            },
            errors: '',
            login: false,
        }

        this.handleLogin = this.handleLogin.bind(this);
        this.handleName = this.handleName.bind(this);
        this.handlePassword = this.handlePassword.bind(this);
        this.handleUser = this.handleUser.bind(this);
    }

    handleUser(name: string){
        let  users  = this.props.users;
        let user = users.find((user: any) => 
            user.name === name || user.email === name
        )
        
        if(user){
            this.setState({
                user: user
            })
        }

    }

    handleLogin(){

            if(this.state.user.password === this.state.password){ 
                this.props.loginUser(this.state.user);
                this.setState({
                    login: true
                })
            }else{
                this.setState({
                    errors: "incorect username or password"
                })
            }
    }

    handleName(event: InputEvent){
        let name = event.target.value
        this.setState({
            name: name
        })
        this.handleUser(name);
    }

    handlePassword(event: InputEvent){
        this.setState({
            password: event.target.value
        })
    }

    render(){
        if(this.state.login){
            return <Redirect to={`/projects/${this.state.user.id}`} />
        }
        return(
            <div>
                <section className="login_form">
                    <legend>Login</legend>
                    <div className="errors">{this.state.errors}</div>
                    <label>Name</label>
                    <input 
                        className="input"
                        type="text" 
                        placeholder="name"
                        onChange={this.handleName}
                        value={this.state.name}
                        />
                    <label>Password</label>                    
                    <input 
                        className="input"
                        type="password" 
                        placeholder="password" 
                        onChange={this.handlePassword}
                        value={this.state.password}
                        />
                    <button 
                    className="button login_button" 
                    type="submit"
                    disabled={!this.state.name && !this.state.password}
                    onClick={this.handleLogin}
                    >Login</button>
                </section>
            </div>
        )
    }
}

const mapDispatchToProps = (dispatch: any) => {
    return {
        loginUser: (user: any)  => dispatch(loginUser(user))
    }
}

const mapStateToProps = (state: any) => {
    return{
        users: state.users,
        user: state.user
    }
}


export default connect(mapStateToProps, mapDispatchToProps)(Login);