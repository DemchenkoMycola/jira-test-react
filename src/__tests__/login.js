import React from 'react';
import Login from '../components/Login'
import renderer from 'react-test-renderer';
import { createStore } from 'redux';

describe('login', () => {

    const store = createStore(() => ({}));

    test('render login form', () =>{
        const login = renderer.create(<Login store={store} />).toJSON();
        expect(login).toMatchSnapshot();
    })

});