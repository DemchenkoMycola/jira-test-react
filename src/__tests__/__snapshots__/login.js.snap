// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`login render login form 1`] = `
<div>
  <section
    className="login_form"
  >
    <legend>
      Login
    </legend>
    <div
      className="errors"
    >
      
    </div>
    <label>
      Name
    </label>
    <input
      className="input"
      onChange={[Function]}
      placeholder="name"
      type="text"
      value=""
    />
    <label>
      Password
    </label>
    <input
      className="input"
      onChange={[Function]}
      placeholder="password"
      type="password"
      value=""
    />
    <button
      className="button"
      disabled={true}
      onClick={[Function]}
      type="submit"
    >
      Login
    </button>
  </section>
</div>
`;
